### PR TITLE
Add models for some Spring pseudo-collections

### PR DESCRIPTION
**Pull request by sauyon**

### CODE REVIEW

File: java/ql/src/semmle/code/java/frameworks/spring/SpringCache.qll [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import java
+
+module SpringCache {
+  private class FlowSummaries extends SummaryModelCsv {
+    override predicate row(string row) {
+      row = [
+        "org.springframework.cache.Cache.ValueRetrievalException;false;getKey;;;MapKey of Argument[-1];ReturnValue;value",
+        "org.springframework.cache.Cache.ValueWrapper;true;get;;;MapValue of Argument[-1];ReturnValue;value",
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringCache.qll:8]: Think you need `$` notation for inner classes

File: java/ql/src/semmle/code/java/frameworks/spring/SpringDataRepository.qll [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    override predicate row(string row) {
+      row = [
+        "org.springframework.data.repository.CrudRepository;true;findAll;;;MapValue of Argument[-1];Element of ReturnValue;value",
+        "org.springframework.data.repository.CrudRepository;true;findAllById;;;MapValue of Argument[-1];Element of ReturnValue;value",
+        // note for review; return value is an Option<T>; assuming Element is used.
+        "org.springframework.data.repository.CrudRepository;true;findById;;;MapValue of Argument[-1];Element of ReturnValue;value",
+        "org.springframework.data.repository.CrudRepository;true;save;;;Argument[0];MapValue of Argument[-1];value",
+        "org.springframework.data.repository.CrudRepository;true;saveAll;;;Argument[0];MapValue of Argument[-1];value",
+        // note for review: is this necessary?
+        "org.springframework.data.repository.PagingAndSortingRepository;true;findAll;;;MapValue of Argument[-1];Element of ReturnValue;value",
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringDataRepository.qll:14]: I think so, as there are not overrides of the superclass (CrudRepository) findAll

**Review by smowton — COMMENTED** [review]
Looks sensible -- I think fully-qualified names might be necessary though; checking

**Comment by smowton** [comment]
Checked: short names do work, but they are always qualified by their generic parameter names if they have them. For example, `(Object,Object)` is fine and so is `(Object,Map<K,V>)` but not `(Object,Map)`. This may not be intentional since it's different to how fully-qualified names are parsed (@aschackmull ?)

**Comment by aschackmull** [comment]
> Checked: short names do work, but they are always qualified by their generic parameter names if they have them. For example, `(Object,Object)` is fine and so is `(Object,Map<K,V>)` but not `(Object,Map)`. This may not be intentional since it's different to how fully-qualified names are parsed (@aschackmull ?)

That's a bug, which is fixed here: https://github.com/github/codeql/pull/5751/commits/f0746e40c19038aaa1eeb7f6709a1509c2c6d6a6

File: java/ql/src/semmle/code/java/frameworks/spring/SpringDataRepository.qll (hunk 2 above)
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringDataRepository.qll:14]:
```suggestion
        "org.springframework.data.repository;CrudRepository;true;findAll;;;MapValue of Argument[-1];Element of ReturnValue;value",
        "org.springframework.data.repository;CrudRepository;true;findAllById;;;MapValue of Argument[-1];Element of ReturnValue;value",
        // note for review; return value is an Option<T>; assuming Element is used.
        "org.springframework.data.repository;CrudRepository;true;findById;;;MapValue of Argument[-1];Element of ReturnValue;value",
        "org.springframework.data.repository;CrudRepository;true;save;;;Argument[0];MapValue of Argument[-1];value",
        "org.springframework.data.repository;CrudRepository;true;saveAll;;;Argument[0];MapValue of Argument[-1];value",
        // note for review: is this necessary?
        "org.springframework.data.repository;PagingAndSortingRepository;true;findAll;;;MapValue of Argument[-1];Element of ReturnValue;value",
```

File: java/ql/src/semmle/code/java/frameworks/spring/SpringCache.qll [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        "org.springframework.cache.Cache;true;get;(Object, Callable<T>);;MapValue of Argument[-1];ReturnValue;value",
+        "org.springframework.cache.Cache;true;get;(Object, Class<T>);;MapValue of Argument[-1];ReturnValue;value",
+        // note for review: there is no guarantee that the result is a map, but all implementations seem to be
+        "org.springframework.cache.Cache;true;getNativeCache;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.cache.Cache;true;getNativeCache;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.cache.Cache;true;put;;;Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.cache.Cache;true;put;;;Argument[1];MapValue of Argument[-1];value",
+        "org.springframework.cache.Cache;true;putIfAbsent;;;Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.cache.Cache;true;putIfAbsent;;;Argument[1];MapValue of Argument[-1];value",
+        "org.springframework.cache.Cache;true;putIfAbsent;;;MapValue of Argument[-1];ReturnValue;value",
```
Comment (smowton) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringCache.qll:19]:
```suggestion
        "org.springframework.cache;Cache.ValueRetrievalException;false;getKey;;;MapKey of Argument[-1];ReturnValue;value",
        "org.springframework.cache;Cache.ValueWrapper;true;get;;;MapValue of Argument[-1];ReturnValue;value",
        "org.springframework.cache;Cache;true;get;(Object);;MapValue of Argument[-1];MapValue of ReturnValue;value",
        "org.springframework.cache;Cache;true;get;(Object, Callable<T>);;MapValue of Argument[-1];ReturnValue;value",
        "org.springframework.cache;Cache;true;get;(Object, Class<T>);;MapValue of Argument[-1];ReturnValue;value",
        // note for review: there is no guarantee that the result is a map, but all implementations seem to be
        "org.springframework.cache;Cache;true;getNativeCache;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
        "org.springframework.cache;Cache;true;getNativeCache;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
        "org.springframework.cache;Cache;true;put;;;Argument[0];MapKey of Argument[-1];value",
        "org.springframework.cache;Cache;true;put;;;Argument[1];MapValue of Argument[-1];value",
        "org.springframework.cache;Cache;true;putIfAbsent;;;Argument[0];MapKey of Argument[-1];value",
        "org.springframework.cache;Cache;true;putIfAbsent;;;Argument[1];MapValue of Argument[-1];value",
        "org.springframework.cache;Cache;true;putIfAbsent;;;MapValue of Argument[-1];ReturnValue;value",
```

File: java/ql/src/semmle/code/java/frameworks/spring/SpringUi.qll [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        "org.springframework.ui;Model;true;addAllAttributes;(Map<String,?>);;MapValue of Argument[0];MapValue of Argument[-1];value",
+        "org.springframework.ui;Model;true;mergeAttributes;(Map<String,?>);;MapKey of Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.ui;Model;true;mergeAttributes;(Map<String,?>);;MapValue of Argument[0];MapValue of Argument[-1];value",
+        "org.springframework.ui;Model;true;getAttribute;;;MapValue of Argument[-1];ReturnValue;value",
+        "org.springframework.ui;Model;true;mergeAttributes;;;MapValue of Argument[-1];ReturnValue;value",
+        "org.springframework.ui;Model;true;asMap;;;MapKey of Argument[-1];MapKey of ReturnValue;value",
+        "org.springframework.ui;Model;true;asMap;;;MapValue of Argument[-1];MapValue of ReturnValue;value",
+        "org.springframework.ui;ModelMap;false;ModelMap;(Object);;Argument[0];MapValue of ReturnValue;value",
+        "org.springframework.ui;ModelMap;false;ModelMap;(String, Object);;Argument[0];MapKey of ReturnValue;value",
+        "org.springframework.ui;ModelMap;false;ModelMap;(String, Object);;Argument[0];MapValue of ReturnValue;value",
```
Comment (atorralba) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringUi.qll:21]:
```suggestion
        "org.springframework.ui;ModelMap;false;ModelMap;(String,Object);;Argument[0];MapKey of ReturnValue;value",
        "org.springframework.ui;ModelMap;false;ModelMap;(String,Object);;Argument[0];MapValue of ReturnValue;value",
```

Last time I checked, spaces were causing the signature to not match. Do you know if this works correctly as it is?

Comment (smowton) [reply java/ql/src/semmle/code/java/frameworks/spring/SpringUi.qll:21]: It doesn't, that's corrected in https://github.com/github/codeql/pull/6030

**Comment by aschackmull** [comment]
Needs conflict resolution.

**Comment by sauyon** [comment]
Done.

File: java/ql/src/semmle/code/java/frameworks/spring/SpringUi.qll [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  override predicate row(string row) {
+    row =
+      [
+        "org.springframework.ui;Model;true;addAllAttributes;;;Argument[-1];ReturnValue;value",
+        "org.springframework.ui;Model;true;addAllAttributes;(Collection);;Element of Argument[0];MapValue of Argument[-1];value",
+        "org.springframework.ui;Model;true;addAllAttributes;(Collection);;Element of Argument[0];MapValue of ReturnValue;value",
+        "org.springframework.ui;Model;true;addAllAttributes;(Map);;MapKey of Argument[0];MapKey of Argument[-1];value",
+        "org.springframework.ui;Model;true;addAllAttributes;(Map);;MapValue of Argument[0];MapValue of Argument[-1];value",
+        "org.springframework.ui;Model;true;addAllAttributes;(Map);;MapKey of Argument[0];MapKey of ReturnValue;value",
+        "org.springframework.ui;Model;true;addAllAttributes;(Map);;MapValue of Argument[0];MapValue of ReturnValue;value",
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/frameworks/spring/SpringUi.qll:18]: I believe these two lines are superfluous - they'll be inferred from the two above in combination with the value step from `Argument[-1]` to `ReturnValue`. Looks like there are also a few more instances of this pattern below.

**Comment by sauyon** [comment]
I removed those models but left the tests in, and they work as expected.

**Review by aschackmull — APPROVED** [review]

**Pull request merged**